src/librbd/ImageCtx.cc

```diff
@@ -803,4 +831,78 @@ namespace librbd {
       async_requests_cond.Wait(async_ops_lock);
     }
   }
+
+  bool ImageCtx::_aware_metadata_confs(const string &prefix, const char **configs, size_t len,
+                                       map<string, bufferlist> &pairs, map<string, bufferlist> *res) {
+    size_t conf_prefix_len = prefix.size();
+
+    string start = prefix;
+    for (map<string, bufferlist>::iterator it = pairs.begin(); it != pairs.end(); ++it) {
+      if (it->first.size() <= conf_prefix_len || it->first.compare(0, conf_prefix_len, prefix))
+        return false;
+
+      for (size_t i = 0; i < len; ++i) {
+        string key = it->first.substr(conf_prefix_len, it->first.size() - conf_prefix_len);
+        if (!key.compare(configs[i])) {
+          res->insert(make_pair(key, it->second));
+          break;
+        }
+      }
+    }
+    return true;
+  }
+
+  void ImageCtx::aware_metadata_confs() {
+    ldout(cct, 20) << __func__ << dendl;
+    static uint64_t max_conf_items = 128;
+
+    string start = METADATA_CONF_PREFIX;
+    int r = 0, j = 0;
+    bool is_continue;
+    do {
+      map<string, bufferlist> pairs, res;
+      r = cls_client::metadata_list(&md_ctx, header_oid, start, max_conf_items, &pairs);
+      if (r < 0) {
+        lderr(cct) << __func__ << " couldn't list conf metadatas: " << r << dendl;
+        break;
+      }
+      if (pairs.empty())
+        break;
+      
+      is_continue = _aware_metadata_confs(METADATA_CONF_PREFIX, AWARE_CONFS, AWARE_CONFS_LEN,
+                                          pairs, &res);
+      for (map<string, bufferlist>::iterator it = res.begin(); it != res.end(); ++it) {
+        j = cct->_conf->set_val(it->first.c_str(), it->second.c_str());
+        if (j < 0)
+          lderr(cct) << __func__ << " failed to set config " << it->first << " with value "
+                     << it->second.c_str() << ": " << j << dendl;
+        break;
+      }
+      start = pairs.rbegin()->first;
+    } while (is_continue);
+
+    cct->_conf->apply_changes(NULL);
```

Comment:
changing the cct->_conf means later images opened using the same cct may have defaults overridden, which doesn't seem like what we want. This'll affect clones, since the parent is opened like a normal image, as well as any librbd users opening more than one image at once with the same rados_t.

Maybe ImageCtx should have its own md_config_t with just the options (not the config_observers) copied from config_t.

Not applying the global settings also means the admin socket config show/get commands won't include these metadata configs, but I'm not sure of a good way to have both, and this seems less important.